hecate_llvm_gen/src/lib.rs

```diff
@@ -0,0 +1,330 @@
+        
+        unsafe {
+            LLVM_InitializeAllTargets();
+            LLVM_InitializeAllTargetInfos();
+            LLVM_InitializeAllAsmParsers();
+            LLVM_InitializeAllAsmPrinters();
+            LLVM_InitializeAllDisassemblers();
+            LLVM_InitializeAllTargetMCs();
+
+            LLVMPrintModuleToFile(self.module, c_str_ptr!(llvm_ir.to_str().unwrap()), std::ptr::null_mut());
+            LLVMWriteBitcodeToFile(self.module, c_str_ptr!(llvm_bc.to_str().unwrap()));
+
+            let triple = LLVMGetDefaultTargetTriple();
+            let features = LLVMGetHostCPUFeatures();
+            let cpu = LLVMGetHostCPUName();
+            let err_string = std::ptr::null_mut();
+            let mut target = std::ptr::null_mut();
+            let target_ret = LLVMGetTargetFromTriple(triple, &mut target, err_string);
+            if target_ret != 0 {
+                panic!("LLVM Error while getting target from triple {}: {}", 
+                    CStr::from_ptr(triple).to_string_lossy(), 
+                    CStr::from_ptr(*err_string).to_string_lossy()
+                );
+            }
+            let target_machine = LLVMCreateTargetMachine(
+                target,
+                triple,
+                cpu,
+                features,
+                match opt {
+                    OptimizationLevel::None => LLVMCodeGenOptLevel::LLVMCodeGenLevelNone,
+                    OptimizationLevel::Less => LLVMCodeGenOptLevel::LLVMCodeGenLevelLess,
+                    OptimizationLevel::Default => LLVMCodeGenOptLevel::LLVMCodeGenLevelDefault,
+                    OptimizationLevel::Aggressive => LLVMCodeGenOptLevel::LLVMCodeGenLevelAggressive,
+                },
+                LLVMRelocMode::LLVMRelocDefault,
+                LLVMCodeModel::LLVMCodeModelDefault,
+            );
+            if target_machine.is_null() {
+                panic!("Could not create LLVM target machine");
+            }
+
+            let err_string = std::ptr::null_mut();
+            let return_code = LLVMTargetMachineEmitToFile(target_machine, self.module, c_str_ptr!(obj_path.to_str().unwrap()) as *mut _, LLVMCodeGenFileType::LLVMObjectFile, err_string);
+            if return_code != 0 {
+                panic!("LLVM Error while compiling to object file: {}", CStr::from_ptr(*err_string).to_string_lossy());
+            }
+
+            let err_string = std::ptr::null_mut();
+            let return_code = LLVMTargetMachineEmitToFile(target_machine, self.module, c_str_ptr!(assembly.to_str().unwrap()) as *mut _, LLVMCodeGenFileType::LLVMAssemblyFile, err_string);
+            if return_code != 0 {
+                panic!("LLVM Error while compiling to object file: {}", CStr::from_ptr(*err_string).to_string_lossy());
+            }
+
+            LLVMDisposeTargetMachine(target_machine);
+        }
+        let r = std::process::Command::new("gcc").arg(obj_path.to_str().unwrap()).arg("-o").arg(exe_path.to_str().unwrap()).output().unwrap();
```

Comment:
Avoid the use of unwraps, especially chained ones. Use `expect`, `unwrap_or` or similar alternatives instead.